// @connext/vector-types/dist/src/schemas/node.js
import { Type } from "@sinclair/typebox";
import { EngineEvents } from "../engine";
import { EngineParams } from "./engine";
import { TUrl, TAddress, TPublicIdentifier, TIntegerString, TBytes32, TFullTransferState, TFullChannelState, TChainId, AllowedSwapSchema, } from "./basic";
const BasicChannelServerResponseSchema = {
    200: Type.Object({
        channelAddress: TAddress,
    }),
};
const BasicTransferServerResponseSchema = {
    200: Type.Object({
        channelAddress: TAddress,
        transferId: TBytes32,
        routingId: Type.Optional(TBytes32),
    }),
};
const GetRouterConfigParamsSchema = Type.Intersect([
    EngineParams.GetRouterConfigSchema,
    Type.Object({
        publicIdentifier: TPublicIdentifier,
    }),
]);
const GetRouterConfigResponseSchema = {
    200: Type.Object({
        supportedChains: Type.Array(TChainId),
        allowedSwaps: Type.Array(AllowedSwapSchema),
    }),
};
const GetTransferStateByRoutingIdParamsSchema = Type.Intersect([
    EngineParams.GetTransferStateByRoutingIdSchema,
    Type.Object({
        publicIdentifier: TPublicIdentifier,
    }),
]);
const GetTransferStateByRoutingIdResponseSchema = {
    200: Type.Union([Type.Undefined(), TFullTransferState]),
};
const GetTransferStatesByRoutingIdParamsSchema = Type.Intersect([
    EngineParams.GetTransferStatesByRoutingIdSchema,
    Type.Object({
        publicIdentifier: TPublicIdentifier,
    }),
]);
const GetTransferStatesByRoutingIdResponseSchema = {
    200: Type.Array(TFullTransferState),
};
const GetActiveTransfersByChannelAddressParamsSchema = Type.Intersect([
    EngineParams.GetActiveTransfersSchema,
    Type.Object({ publicIdentifier: TPublicIdentifier }),
]);
const GetActiveTransfersByChannelAddressResponseSchema = {
    200: Type.Array(TFullTransferState),
};
const GetTransferStateParamsSchema = Type.Intersect([
    EngineParams.GetTransferStateSchema,
    Type.Object({ publicIdentifier: TPublicIdentifier }),
]);
const GetTransferStateResponseSchema = {
    200: Type.Union([Type.Undefined(), TFullTransferState]),
};
const GetChannelStateParamsSchema = Type.Intersect([
    EngineParams.GetChannelStateSchema,
    Type.Object({ publicIdentifier: TPublicIdentifier }),
]);
const GetChannelStateResponseSchema = {
    200: Type.Union([Type.Undefined(), TFullChannelState]),
};
const GetChannelStatesParamsSchema = Type.Intersect([
    EngineParams.GetChannelStatesSchema,
    Type.Object({ publicIdentifier: TPublicIdentifier }),
]);
const GetChannelStatesResponseSchema = {
    200: Type.Array(TAddress),
};
const GetChannelStateByParticipantsParamsSchema = Type.Object({
    publicIdentifier: TPublicIdentifier,
    counterparty: TPublicIdentifier,
    chainId: TChainId,
});
const GetChannelStateByParticipantsResponseSchema = GetChannelStateResponseSchema;
const GetConfigResponseSchema = {
    200: Type.Array(Type.Object({
        publicIdentifier: TPublicIdentifier,
        signerAddress: TAddress,
        index: Type.Integer(),
    })),
};
const GetStatusResponseSchema = {
    200: Type.Object({
        publicIdentifier: TPublicIdentifier,
        signerAddress: TAddress,
        providerSyncing: Type.Dict(Type.Union([
            Type.Boolean(),
            Type.Object({
                startingBlock: Type.String(),
                currentBlock: Type.String(),
                highestBlock: Type.String(),
            }),
            Type.String(),
            Type.Undefined(),
        ])),
        version: Type.String(),
    }),
};
const GetListenerParamsSchema = Type.Object({
    eventName: Type.Union(Object.values(EngineEvents).map((e) => Type.Literal(e))),
    publicIdentifier: TPublicIdentifier,
});
const GetListenerResponseSchema = {
    200: Type.Object({ url: TUrl }),
};
const GetListenersParamsSchema = Type.Object({ publicIdentifier: TPublicIdentifier });
const GetListenersResponseSchema = {
    200: Type.Dict(TUrl),
};
const GetRegisteredTransfersParamsSchema = Type.Intersect([
    EngineParams.GetRegisteredTransfersSchema,
    Type.Object({ publicIdentifier: TPublicIdentifier }),
]);
const GetRegisteredTransfersResponseSchema = {
    200: Type.Array(Type.Object({
        name: Type.String(),
        stateEncoding: Type.String(),
        resolverEncoding: Type.String(),
        definition: TAddress,
        encodedCancel: Type.String(),
    })),
};
const PostRegisterListenerBodySchema = Type.Object({
    publicIdentifier: TPublicIdentifier,
    events: Type.Dict(Type.String()),
});
const PostRegisterListenerResponseSchema = {
    200: Type.Object({
        message: Type.String(),
    }),
};
const PostSetupBodySchema = Type.Intersect([
    EngineParams.SetupSchema,
    Type.Object({ publicIdentifier: TPublicIdentifier }),
]);
const PostSetupResponseSchema = BasicChannelServerResponseSchema;
const PostRequestSetupBodySchema = PostSetupBodySchema;
const PostRequestSetupResponseSchema = BasicChannelServerResponseSchema;
const PostDepositBodySchema = Type.Intersect([
    EngineParams.DepositSchema,
    Type.Object({ publicIdentifier: TPublicIdentifier }),
]);
const PostDepositResponseSchema = BasicChannelServerResponseSchema;
const PostRequestCollateralBodySchema = Type.Intersect([
    EngineParams.RequestCollateralSchema,
    Type.Object({ publicIdentifier: TPublicIdentifier }),
]);
const PostRequestCollateralResponseSchema = BasicChannelServerResponseSchema;
const PostSendDepositTxBodySchema = Type.Object({
    channelAddress: TAddress,
    amount: TIntegerString,
    assetId: TAddress,
    chainId: TChainId,
    publicIdentifier: TPublicIdentifier,
});
const PostSendDepositTxResponseSchema = {
    200: Type.Object({
        txHash: TBytes32,
    }),
};
const PostConditionalTransferBodySchema = Type.Intersect([
    EngineParams.ConditionalTransferSchema,
    Type.Object({ publicIdentifier: TPublicIdentifier }),
]);
const PostConditionalTransferResponseSchema = BasicTransferServerResponseSchema;
const PostResolveTransferBodySchema = Type.Intersect([
    EngineParams.ResolveTransferSchema,
    Type.Object({ publicIdentifier: TPublicIdentifier }),
]);
const PostResolveTransferResponseSchema = BasicTransferServerResponseSchema;
const PostWithdrawTransferBodySchema = Type.Intersect([
    EngineParams.WithdrawSchema,
    Type.Object({ publicIdentifier: TPublicIdentifier }),
]);
const PostWithdrawTransferResponseSchema = {
    200: Type.Object({
        channelAddress: TAddress,
        transferId: TBytes32,
        transactionHash: Type.Optional(TBytes32),
    }),
};
const PostSignUtilityMessageBodySchema = Type.Intersect([
    EngineParams.SignUtilityMessageSchema,
    Type.Object({ publicIdentifier: TPublicIdentifier }),
]);
const PostSignUtilityMessageResponseSchema = {
    200: Type.Object({
        signedMessage: Type.String(),
    }),
};
const PostRestoreStateBodySchema = Type.Intersect([
    EngineParams.RestoreStateSchema,
    Type.Object({ publicIdentifier: TPublicIdentifier }),
]);
const PostRestoreStateResponseSchema = {
    200: Type.Object({
        channelAddress: TAddress,
    }),
};
const PostCreateNodeBodySchema = Type.Object({
    index: Type.Integer({ minimum: 0, maximum: 2147483647 }),
    mnemonic: Type.Optional(Type.String()),
    skipCheckIn: Type.Optional(Type.Boolean()),
});
const PostCreateNodeResponseSchema = {
    200: Type.Object({
        publicIdentifier: TPublicIdentifier,
        signerAddress: TAddress,
        index: Type.Integer(),
    }),
};
const PostAdminBodySchema = Type.Object({
    adminToken: Type.String({
        example: "cxt1234",
        description: "Admin token",
    }),
});
const PostAdminResponseSchema = {
    200: Type.Object({
        message: Type.String(),
    }),
};
const PostAdminRetryWithdrawTransactionBodySchema = Type.Object({
    adminToken: Type.String(),
    transferId: TBytes32,
});
const PostAdminRetryWithdrawTransactionResponseSchema = {
    200: Type.Object({
        transferId: TBytes32,
        transactionHash: Type.String(),
    }),
};
const PostSendDisputeChannelTxBodySchema = Type.Intersect([
    EngineParams.DisputeChannelSchema,
    Type.Object({ publicIdentifier: TPublicIdentifier }),
]);
const PostSendDisputeChannelTxResponseSchema = {
    200: Type.Object({
        txHash: TBytes32,
    }),
};
const PostSendDefundChannelTxBodySchema = Type.Intersect([
    EngineParams.DefundChannelSchema,
    Type.Object({ publicIdentifier: TPublicIdentifier }),
]);
const PostSendDefundChannelTxResponseSchema = {
    200: Type.Object({
        txHash: TBytes32,
    }),
};
const PostSendDisputeTransferTxBodySchema = Type.Intersect([
    EngineParams.DisputeTransferSchema,
    Type.Object({ publicIdentifier: TPublicIdentifier }),
]);
const PostSendDisputeTransferTxResponseSchema = {
    200: Type.Object({
        txHash: TBytes32,
    }),
};
const PostSendDefundTransferTxBodySchema = Type.Intersect([
    EngineParams.DefundTransferSchema,
    Type.Object({ publicIdentifier: TPublicIdentifier }),
]);
const PostSendDefundTransferTxResponseSchema = {
    200: Type.Object({
        txHash: TBytes32,
    }),
};
const PostSendIsAliveBodySchema = Type.Intersect([
    EngineParams.SendIsAliveSchema,
    Type.Object({ publicIdentifier: TPublicIdentifier }),
]);
const PostSendIsAliveResponseSchema = {
    200: Type.Object({
        channelAddress: TAddress,
    }),
};
export var NodeParams;
(function (NodeParams) {
    NodeParams.GetRouterConfigSchema = GetRouterConfigParamsSchema;
    NodeParams.GetTransferStateByRoutingIdSchema = GetTransferStateByRoutingIdParamsSchema;
    NodeParams.GetTransferStatesByRoutingIdSchema = GetTransferStatesByRoutingIdParamsSchema;
    NodeParams.GetTransferStateSchema = GetTransferStateParamsSchema;
    NodeParams.GetActiveTransfersByChannelAddressSchema = GetActiveTransfersByChannelAddressParamsSchema;
    NodeParams.GetChannelStateSchema = GetChannelStateParamsSchema;
    NodeParams.GetChannelStatesSchema = GetChannelStatesParamsSchema;
    NodeParams.GetChannelStateByParticipantsSchema = GetChannelStateByParticipantsParamsSchema;
    NodeParams.GetListenerSchema = GetListenerParamsSchema;
    NodeParams.GetListenersSchema = GetListenersParamsSchema;
    NodeParams.GetRegisteredTransfersSchema = GetRegisteredTransfersParamsSchema;
    NodeParams.GetConfigSchema = Type.Object({});
    NodeParams.SetupSchema = PostSetupBodySchema;
    NodeParams.RequestSetupSchema = PostRequestSetupBodySchema;
    NodeParams.DepositSchema = PostDepositBodySchema;
    NodeParams.RequestCollateralSchema = PostRequestCollateralBodySchema;
    NodeParams.SendDepositTxSchema = PostSendDepositTxBodySchema;
    NodeParams.ConditionalTransferSchema = PostConditionalTransferBodySchema;
    NodeParams.ResolveTransferSchema = PostResolveTransferBodySchema;
    NodeParams.WithdrawSchema = PostWithdrawTransferBodySchema;
    NodeParams.RegisterListenerSchema = PostRegisterListenerBodySchema;
    NodeParams.SignUtilityMessageSchema = PostSignUtilityMessageBodySchema;
    NodeParams.RestoreStateSchema = PostRestoreStateBodySchema;
    NodeParams.AdminSchema = PostAdminBodySchema;
    NodeParams.CreateNodeSchema = PostCreateNodeBodySchema;
    NodeParams.SendDisputeChannelTxSchema = PostSendDisputeChannelTxBodySchema;
    NodeParams.SendDefundChannelTxSchema = PostSendDefundChannelTxBodySchema;
    NodeParams.SendDisputeTransferTxSchema = PostSendDisputeTransferTxBodySchema;
    NodeParams.SendDefundTransferTxSchema = PostSendDefundTransferTxBodySchema;
    NodeParams.SendIsAliveSchema = PostSendIsAliveBodySchema;
    NodeParams.RetryWithdrawTransactionSchema = PostAdminRetryWithdrawTransactionBodySchema;
})(NodeParams || (NodeParams = {}));
export var NodeResponses;
(function (NodeResponses) {
    NodeResponses.GetRouterConfigSchema = GetRouterConfigResponseSchema;
    NodeResponses.GetTransferStateByRoutingIdSchema = GetTransferStateByRoutingIdResponseSchema;
    NodeResponses.GetTransferStatesByRoutingIdSchema = GetTransferStatesByRoutingIdResponseSchema;
    NodeResponses.GetTransferStateSchema = GetTransferStateResponseSchema;
    NodeResponses.GetActiveTransfersByChannelAddressSchema = GetActiveTransfersByChannelAddressResponseSchema;
    NodeResponses.GetChannelStateSchema = GetChannelStateResponseSchema;
    NodeResponses.GetChannelStateByParticipantsSchema = GetChannelStateByParticipantsResponseSchema;
    NodeResponses.GetChannelStatesSchema = GetChannelStatesResponseSchema;
    NodeResponses.GetListenerSchema = GetListenerResponseSchema;
    NodeResponses.GetListenersSchema = GetListenersResponseSchema;
    NodeResponses.GetConfigSchema = GetConfigResponseSchema;
    NodeResponses.GetStatusSchema = GetStatusResponseSchema;
    NodeResponses.GetRegisteredTransfersSchema = GetRegisteredTransfersResponseSchema;
    NodeResponses.SetupSchema = PostSetupResponseSchema;
    NodeResponses.RequestSetupSchema = PostRequestSetupResponseSchema;
    NodeResponses.DepositSchema = PostDepositResponseSchema;
    NodeResponses.RequestCollateralSchema = PostRequestCollateralResponseSchema;
    NodeResponses.SendDepositTxSchema = PostSendDepositTxResponseSchema;
    NodeResponses.ConditionalTransferSchema = PostConditionalTransferResponseSchema;
    NodeResponses.ResolveTransferSchema = PostResolveTransferResponseSchema;
    NodeResponses.WithdrawSchema = PostWithdrawTransferResponseSchema;
    NodeResponses.RegisterListenerSchema = PostRegisterListenerResponseSchema;
    NodeResponses.SignUtilityMessageSchema = PostSignUtilityMessageResponseSchema;
    NodeResponses.RestoreStateSchema = PostRestoreStateResponseSchema;
    NodeResponses.AdminSchema = PostAdminResponseSchema;
    NodeResponses.CreateNodeSchema = PostCreateNodeResponseSchema;
    NodeResponses.SendDisputeChannelTxSchema = PostSendDisputeChannelTxResponseSchema;
    NodeResponses.SendDefundChannelTxSchema = PostSendDefundChannelTxResponseSchema;
    NodeResponses.SendDisputeTransferTxSchema = PostSendDisputeTransferTxResponseSchema;
    NodeResponses.SendDefundTransferTxSchema = PostSendDefundTransferTxResponseSchema;
    NodeResponses.SendIsAliveSchema = PostSendIsAliveResponseSchema;
    NodeResponses.RetryWithdrawTransactionSchema = PostAdminRetryWithdrawTransactionResponseSchema;
})(NodeResponses || (NodeResponses = {}));
