// @connext/vector-types/dist/src/engine.js
export const IS_ALIVE_EVENT = "IS_ALIVE";
export const SETUP_EVENT = "SETUP";
export const CONDITIONAL_TRANSFER_CREATED_EVENT = "CONDITIONAL_TRANSFER_CREATED";
export const CONDITIONAL_TRANSFER_RESOLVED_EVENT = "CONDITIONAL_TRANSFER_RESOLVED";
export const DEPOSIT_RECONCILED_EVENT = "DEPOSIT_RECONCILED";
export const REQUEST_COLLATERAL_EVENT = "REQUEST_COLLATERAL";
export const WITHDRAWAL_CREATED_EVENT = "WITHDRAWAL_CREATED";
export const WITHDRAWAL_RESOLVED_EVENT = "WITHDRAWAL_RESOLVED";
export const WITHDRAWAL_RECONCILED_EVENT = "WITHDRAWAL_RECONCILED";
export const RESTORE_STATE_EVENT = "RESTORE_STATE_EVENT";
export const EngineEvents = {
    [IS_ALIVE_EVENT]: IS_ALIVE_EVENT,
    [SETUP_EVENT]: SETUP_EVENT,
    [CONDITIONAL_TRANSFER_CREATED_EVENT]: CONDITIONAL_TRANSFER_CREATED_EVENT,
    [CONDITIONAL_TRANSFER_RESOLVED_EVENT]: CONDITIONAL_TRANSFER_RESOLVED_EVENT,
    [DEPOSIT_RECONCILED_EVENT]: DEPOSIT_RECONCILED_EVENT,
    [REQUEST_COLLATERAL_EVENT]: REQUEST_COLLATERAL_EVENT,
    [RESTORE_STATE_EVENT]: RESTORE_STATE_EVENT,
    [SETUP_EVENT]: SETUP_EVENT,
    [WITHDRAWAL_CREATED_EVENT]: WITHDRAWAL_CREATED_EVENT,
    [WITHDRAWAL_RESOLVED_EVENT]: WITHDRAWAL_RESOLVED_EVENT,
    [WITHDRAWAL_RECONCILED_EVENT]: WITHDRAWAL_RECONCILED_EVENT,
};
