// @connext/vector-types/dist/src/vectorProvider.js
export const ChannelRpcMethods = {
    chan_signUtilityMessage: "chan_signUtilityMessage",
    chan_getConfig: "chan_getConfig",
    chan_getRouterConfig: "chan_getRouterConfig",
    chan_getStatus: "chan_getStatus",
    chan_getChannelState: "chan_getChannelState",
    chan_getChannelStateByParticipants: "chan_getChannelStateByParticipants",
    chan_getChannelStates: "chan_getChannelStates",
    chan_getTransferStateByRoutingId: "chan_getTransferStateByRoutingId",
    chan_getTransferStatesByRoutingId: "chan_getTransferStatesByRoutingId",
    chan_getActiveTransfers: "chan_getActiveTransfers",
    chan_getRegisteredTransfers: "chan_getRegisteredTransfers",
    chan_getTransferState: "chan_getTransferState",
    chan_setup: "chan_setup",
    chan_sendIsAlive: "chan_sendIsAlive",
    chan_requestSetup: "chan_requestSetup",
    chan_deposit: "chan_deposit",
    chan_requestCollateral: "chan_requestCollateral",
    chan_createTransfer: "chan_createTransfer",
    chan_resolveTransfer: "chan_resolveTransfer",
    chan_restoreState: "chan_restoreState",
    chan_withdraw: "chan_withdraw",
    chan_subscribe: "chan_subscribe",
    chan_unsubscribeAll: "chan_unsubscribeAll",
    connext_authenticate: "connext_authenticate",
    chan_dispute: "chan_dispute",
    chan_defund: "chan_defund",
    chan_disputeTransfer: "chan_disputeTransfer",
    chan_defundTransfer: "chan_defundTransfer",
    chan_decrypt: "chan_decrypt",
    chan_subscription: "chan_subscription",
};
